parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,351 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+import os
+from functools import lru_cache
+
+import torch
+from torch import nn
+
+from parlai.core.torch_agent import TorchAgent, Output
+from parlai.core.thread_utils import SharedTable
+from parlai.core.utils import round_sigfigs, padded_tensor
+
+class TorchRankerAgent(TorchAgent):
+    @staticmethod
+    def add_cmdline_args(argparser):
+        TorchAgent.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('TorchRankerAgent')
+        agent.add_argument(
+            '-tc', '--train_candidates', type=str, default='batch',
+            choices=['batch', 'inline', 'fixed', 'dict'],
+            help='the source of candidates during training')
+        agent.add_argument(
+            '-ec', '--eval_candidates', type=str, default='batch',
+            choices=['batch', 'inline', 'fixed', 'dict'],
```

Comment:
default to fixed?

Reply:
So, I was torn on this. Defaulting to fixed seems a little scary to me because 'fixed' here means a single identical candidate set for all examples (e.g., when in interactive mode). So we'll want something different during training anyway. And not all tasks have a fixed candidate set, so our default would be something that's not always applicable, (possibly resulting in a person using whatever we backed off to when they think they're using fixed), whereas you'll always have a batch. Thoughts?

Reply:
perhaps `inline` is the behavior we want? If the dataset provides a set of `label_candidates`, the default should be to use them.